drv/lpc55-flash/src/lib.rs

```diff
@@ -0,0 +1,370 @@
+//! A raw driver for the LPC55 flash controller without using the ROM.
+//!
+//! This driver is written in a very generic form that doesn't assume any
+//! particular execution model. You can use it to implement busy waiting, or
+//! an interrupt driven driver, etc.
+//!
+//! See the [`Flash`] type for more details.
+
+#![no_std]
+
+use core::ops::RangeInclusive;
+
+/// Number of bytes per flash word.
+pub const BYTES_PER_FLASH_WORD: usize = 16;
+
+/// Number of bytes per flash page.
+pub const BYTES_PER_FLASH_PAGE: usize = 512;
+
+/// Flash driver handle. Wraps a pointer to the flash register set and provides
+/// encapsulated operations.
+///
+/// # Procedure for writing to flash
+///
+/// 1. Use `start_erase_region` to begin an erase of your desired region. Use
+///    `poll_erase_or_program_result` to discover when it's done. (Note: you can
+///    use `start_blank_check` to skip this step if there's a chance the flash
+///    is already erased.)
+/// 2. Call `start_write_row` with each 16 byte chunk of the first page you're
+///    writing, 32 times, for rows 0-31. Each time, use `poll_write_result` to
+///    check that it's done.
+/// 3. Call `start_program` with the index of any word in the target flash page
```

Comment:
This is kind of wacky to me. That you only have to pick one of the words to perform the programming, and it doesn't matter which one.

Reply:
Yeah, this is me exposing how the hardware works, but it might not be the best idea.

Erase and programming (and blank check) all affect pages, but take word indices as their inputs. They ignore the bottom bits. So if you ask to erase word W you're actually going to erase the page containing word W. The API currently exposes this.

I could have page operations take page indices instead. I didn't do this in the first pass because
1. My default is to expose the hardware's behavior so that you can compare it to the datasheet, and
2. This would be the first instance of a page index in the API, which is currently all word index based. Having two different types of addresses is _also_ potentially error prone.

But that could be wrong! What do you think? Also @mkeeter

Reply:
I'm okay with the current behavior, because I agree that matching the datasheet is valuable.

Reply:
Yeah, synced with @andrewjstone in chat and he's on the same page. I have added comments clarifying the reason for the behavior.